logger/logger.go

```diff
@@ -4,35 +4,104 @@ package logger
 import (
 	"fmt"
 	"io"
+	"io/ioutil"
 	"log"
 	"os"
 	"path/filepath"
 	"time"
 )
 
+// LogOption enable to customize Logger
+type LogOption func(l *Logger) error
+
 var loggingFiles []io.WriteCloser
 
 // Logger has information for logging
 type Logger struct {
-	File  io.WriteCloser
-	Debug *log.Logger // Debug is for debug log. You can set log flag. Also you can choose log stderr or not
-	Error *log.Logger // Error is always debug file and stderr with LstdFlags flag.
+	debugFlag         int
+	errorFlag         int
+	hasOutputDebugLog bool
+	file              io.WriteCloser
+	Debug             *log.Logger // Debug is for debug log. You can set log flag. Also you can choose log stderr or not
+	Error             *log.Logger // Error is always debug file and stderr with LstdFlags flag.
+}
+
+// DebugFlag return current log debugFlag status
+func (l *Logger) DebugFlag() int {
+	return l.debugFlag
+}
+
+// ErrorFlag return current log errorFlag status
+func (l *Logger) ErrorFlag() int {
+	return l.errorFlag
+}
+
+// File return current log output file. If the file = nil, logger does not output log to file
+func (l *Logger) File() io.WriteCloser {
+	return l.file
+}
+
+// OutputToFileWithCreate create file for output log
+func OutputToFileWithCreate(dir, filename string) LogOption {
+	return func(l *Logger) error {
+		file, err := createLogFile(dir, filename)
+		if err != nil {
+			return err
+		}
+		l.file = file
+		return nil
+	}
+}
+
+// OutputToFile output log to the file
+func OutputToFile(file io.WriteCloser) LogOption {
+	return func(l *Logger) error {
+		l.file = file
+		return nil
+	}
+}
+
+// DebugFlag set Logger.Debug flag
+func DebugFlag(flag int) LogOption {
+	return func(l *Logger) error {
+		l.debugFlag = flag
+		return nil
+	}
+}
+
+// ErrorFlag set Logger.Error flag
+func ErrorFlag(flag int) LogOption {
+	return func(l *Logger) error {
+		l.errorFlag = flag
+		return nil
+	}
+}
+
+// OutputDebugLog output debug log
+func OutputDebugLog(output bool) LogOption {
+	return func(l *Logger) error {
+		l.hasOutputDebugLog = output
+		return nil
+	}
 }
 
 // New returns logger object
-func New(dirPath, filename string, flag int, debug bool) (*Logger, error) {
+func New(options ...LogOption) (*Logger, error) {
 	lgr := new(Logger)
+	lgr.errorFlag = log.LstdFlags
```

Comment:
Is it better to set the default for `lgr.debugFlag` here too?
Then the following functions could be made test-only functions.
- `func DebugFlag(flag int) LogOption`
- `func ErrorFlag(flag int) LogOption`

Reply:
You are right. I fixed

Reply:
Sorry, I forget the second opinion.
I think it's ok to put LogOption function to non-test code, because it it not strange that logger package have these Option.
Also, if I write these Option to test code, it is hard to set debug option from executor test.

Reply:
Thx. I see.